pkg/volume/flocker/plugin.go

```diff
@@ -0,0 +1,229 @@
+	host := getenvOrFallback("FLOCKER_CONTROL_SERVICE_HOST", defaultHost)
+	portConfig := getenvOrFallback("FLOCKER_CONTROL_SERVICE_PORT", strconv.Itoa(defaultPort))
+	port, err := strconv.Atoi(portConfig)
+	if err != nil {
+		return nil, err
+	}
+	caCertPath := getenvOrFallback("FLOCKER_CONTROL_SERVICE_CA_FILE", defaultCACertFile)
+	keyPath := getenvOrFallback("FLOCKER_CONTROL_SERVICE_CLIENT_KEY_FILE", defaultClientKeyFile)
+	certPath := getenvOrFallback("FLOCKER_CONTROL_SERVICE_CLIENT_CERT_FILE", defaultClientCertFile)
+
+	c, err := flockerClient.NewClient(host, port, b.flocker.pod.Status.HostIP, caCertPath, keyPath, certPath)
+	return c, err
+}
+
+func (b *flockerBuilder) getMetaDir() string {
+	return path.Join(
+		b.plugin.host.GetPodPluginDir(
+			b.flocker.pod.UID, util.EscapeQualifiedNameForDisk(flockerPluginName),
+		),
+		b.datasetName,
+	)
+}
+
+/*
+SetUpAt will setup a Flocker volume following this flow of calls to the Flocker
+control service:
+
+1. Get the dataset id for the given volume name/dir
+2. It should already be there, if it's not the user needs to manually create it
+3. Check the current Primary UUID
+4. If it doesn't match with the Primary UUID that we got on 2, then we will
+   need to update the Primary UUID for this volume.
+5. Wait until the Primary UUID was updated or timeout.
+*/
+func (b flockerBuilder) SetUpAt(dir string) error {
+	if volumeutil.IsReady(b.getMetaDir()) {
+		return nil
+	}
+
+	if b.client == nil {
+		c, err := b.newFlockerClient()
+		if err != nil {
+			return err
+		}
+		b.client = c
+	}
+
+	datasetID, err := b.client.GetDatasetID(dir)
+	if err != nil {
+		return err
+	}
+
+	s, err := b.client.GetDatasetState(datasetID)
+	if err != nil {
+		return fmt.Errorf("The volume '%s' is not available in Flocker. You need to create this manually with Flocker CLI before using it.", dir)
+	}
+
+	primaryUUID, err := b.client.GetPrimaryUUID()
+	if err != nil {
+		return err
+	}
+
+	if s.Primary != primaryUUID {
+		if err := b.updateDatasetPrimary(datasetID, primaryUUID); err != nil {
+			return err
+		}
+	}
+
+	b.flocker.path = s.Path
+	volumeutil.SetReady(b.getMetaDir())
+	return nil
+}
+
+func (b flockerBuilder) IsReadOnly() bool {
+	return b.readOnly
+}
+
+// updateDatasetPrimary will update the primary in Flocker and wait for it to
+// be ready. If it never gets to ready state it will timeout and error.
+func (b flockerBuilder) updateDatasetPrimary(datasetID, primaryUUID string) error {
+	// We need to update the primary and wait for it to be ready
+	_, err := b.client.UpdatePrimaryForDataset(primaryUUID, datasetID)
+	if err != nil {
+		return err
+	}
+
+	timeoutChan := time.NewTimer(timeoutWaitingForVolume).C
+	tickChan := time.NewTicker(tickerWaitingForVolume).C
+
+	for {
+		if s, err := b.client.GetDatasetState(datasetID); err == nil && s.Primary == primaryUUID {
+			return nil
+		}
+
+		select {
+		case <-timeoutChan:
+			return err
```

Comment:
nit, but I'd still like to see this addressed:

There is a possibility that `err` will be nil, so please wrap the `err`: instead of `return err` do `return fmt.Errorf("Timed out waiting for... %v", err)`

I'm a fan of defensive programming.